internal/slicer/slicer.go

```diff
@@ -151,7 +159,7 @@ func Run(options *RunOptions) (*Report, error) {
 		}
 		reader, err := archives[slice.Package].Fetch(slice.Package)
 		if err != nil {
-			return nil, err
+			return err
 		}
 		defer reader.Close()
 		packages[slice.Package] = reader
```

Comment:
At this moment, we've just made a search through the archive index, fiddled with the fields in there, and then handed it to a fetcher to download the file.  That means we have all that data we need for the package info at hand. It would be cheaper to simply return data from the fetcher a second parameter, than to do a second iteration over all packages later.

If we had the manifest at hand here, we wouldn't even need to be storing yet another intermediate state in memory: we already have the actual index file, the manifest, and now we'd be adding a map of package name to package details, a third time.

Reply:
Done in https://github.com/canonical/chisel/pull/142/commits/687b21e3e35a744f201e8c0c2fbcd9a348d45f51.